![](logo1.jpg)

# **shAI Training 2023 | Level 1**

## Task #8 (End-to-End ML Project {part_2})

## Welcome to the exercises for reviewing second part of end to end ML project.
**Make sure that you read and understand ch2 from the hands-on ML book (page 72 to the end of the chapter ) before start with this notebook.**

**If you stuck with anything reread that part from the book and feel free to ask about anything in the messenger group as you go along.**

 ## Good Luck : )

## first run the following cell for the first part of the project to continue your work 

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline 
import seaborn as sns
from sklearn.model_selection import train_test_split
from pandas.plotting import scatter_matrix
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer

In [45]:
import os
import tarfile
import urllib
DOWNLOAD_ROOT = "https://raw.githubusercontent.com/ageron/handson-ml2/master/"
HOUSING_PATH = os.path.join("datasets", "housing")
HOUSING_URL = DOWNLOAD_ROOT + "datasets/housing/housing.tgz"

def fetch_housing_data(housing_url=HOUSING_URL, housing_path=HOUSING_PATH):
    os.makedirs(housing_path, exist_ok=True)
    tgz_path = os.path.join(housing_path, "housing.tgz")
    urllib.request.urlretrieve(housing_url, tgz_path)
    housing_tgz = tarfile.open(tgz_path)
    housing_tgz.extractall(path=housing_path)
    housing_tgz.close()
    
def load_housing_data(housing_path=HOUSING_PATH):
   csv_path = os.path.join(housing_path, "housing.csv")
   return pd.read_csv(csv_path)
   
fetch_housing_data()
housing = load_housing_data()

rooms_ix, bedrooms_ix, population_ix, household_ix = [
    list(housing.columns).index(col)
    for col in ("total_rooms", "total_bedrooms", "population", "households")]

class CombinedAttributesAdder(BaseEstimator, TransformerMixin):
    def __init__(self, add_bedrooms_per_room = True):
        self.add_bedrooms_per_room = add_bedrooms_per_room
    def fit(self, X, y=None):
        return self  # nothing else to do
    def transform(self, X, y=None):
        rooms_per_household = X[:, rooms_ix] / X[:, household_ix]
        population_per_household = X[:, population_ix] / X[:, household_ix]
        if self.add_bedrooms_per_room:
            bedrooms_per_room = X[:, bedrooms_ix] / X[:, rooms_ix]
            return np.c_[X, rooms_per_household, population_per_household,
                         bedrooms_per_room]
        else:
            return np.c_[X, rooms_per_household, population_per_household]
        
train_set, test_set = train_test_split(housing, test_size=0.2, random_state=42)
housing = train_set.drop("median_house_value", axis=1)
housing_labels = train_set["median_house_value"].copy()

housing_num = housing.drop("ocean_proximity", axis=1)
num_attribs = list(housing_num)
cat_attribs = ["ocean_proximity"]

num_pipeline = Pipeline([
 ('imputer', SimpleImputer(strategy="median")),
 ('attribs_adder', CombinedAttributesAdder()),
 ('std_scaler', StandardScaler())])

full_pipeline = ColumnTransformer([
 ("num", num_pipeline, num_attribs),
 ("cat", OneHotEncoder(), cat_attribs)])

housing_prepared = full_pipeline.fit_transform(housing)

# 1- Select and Train a Model

# Let’s first train a LinearRegression model 

In [36]:
# CODE HERE
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(housing_prepared, housing_labels)

LinearRegression()

# First try it out on a few instances from the training set:


In [49]:
some_data = housing.iloc[:5]
some_data = full_pipeline.transform(some_data)
some_labels = housing_labels.iloc[:5]

In [52]:
# CODE HERE
predictions =  lin_reg.predict(some_data)
predictions.round(-2)

array([181700., 290600., 245000., 146500., 163200.])

In [31]:
some_labels

14196    103000.0
8267     382100.0
17445    172600.0
14265     93400.0
2271      96500.0
Name: median_house_value, dtype: float64

# measure this regression model’s RMSE on the whole training set 
* sing Scikit-Learn’s mean_squared_error() function:

In [53]:
from sklearn.metrics import mean_squared_error

In [71]:
# CODE HERE
predicted = lin_reg.predict(housing_prepared)
lin_rmse = mean_squared_error(housing_labels, predicted, squared=False)
lin_rmse

/home/khalid/.local/lib/python3.10/site-packages/sklearn/metrics/_regression.py:483: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


67593.20745775253

# judge on the RMSE result for this model 
write down your answar 

your answer goes here:

clearly not a great score: most districts median_housing_values range between {$120,000} and $265,000, so a
typical prediction error of $67,593 is really not very satisfying. This is an
example of a model underfitting the training data. When this happens it can
mean that the features do not provide enough information to make good
predictions, or that the model is not powerful enough.

# Let’s train a Decision Tree Regressor model 
## more powerful model

In [56]:
from sklearn.tree import DecisionTreeRegressor 

In [57]:
# CODE HERE
decison_tree_model = DecisionTreeRegressor()
decison_tree_model.fit(housing_prepared, housing_labels)

DecisionTreeRegressor()

# Now evaluate the model on the training set 
* using Scikit-Learn’s mean_squared_error() function:

In [58]:
# CODE HERE
predicted = decison_tree_model.predict(housing_prepared)
mean_squared_error(housing_labels, predicted)

0.0

# Explaine this result 
write down your answar

your answer goes here
it is much more likely that the model has badly overfit the data

# Evaluation Using Cross-Validation

1-split the training set into 10 distinct subsets then train and evaluate the Decision Tree model

In [62]:
from sklearn.model_selection import cross_val_score

In [64]:
# CODE HERE
tree_rmses = -cross_val_score(decison_tree_model, housing_prepared, housing_labels,
scoring="neg_root_mean_squared_error",
cv=10)

2- display the resultant scores and calculate its Mean and Standard deviation

In [65]:
# CODE HERE
pd.Series(tree_rmses).describe()

count       10.000000
mean     68622.095894
std       2655.934630
min      64397.927591
25%      67059.063469
50%      68254.568994
75%      70890.022846
max      72591.107633
dtype: float64

3-repaet the same steps to compute the same scores for the Linear Regression  model 

*notice the difference between the results of the two models*

In [66]:
# CODE HERE
lin_reg_rmses = -cross_val_score(lin_reg, housing_prepared, housing_labels,
scoring="neg_root_mean_squared_error",
cv=10)
pd.Series(lin_reg_rmses).describe()


count       10.000000
mean     67828.386774
std       2601.596761
min      65000.673826
25%      65472.168399
50%      67762.593108
75%      68849.373294
max      72739.875560
dtype: float64

## Let’s train one last model the RandomForestRegressor.

In [67]:
# CODE HERE
from sklearn.ensemble import RandomForestRegressor
forest_reg =  RandomForestRegressor(random_state=42)

# repeat the same steps to compute the same scores its Mean and Standard deviation for the Random Forest model

In [68]:
# CODE HERE
forest_rmses = -cross_val_score(forest_reg, housing_prepared, housing_labels,
scoring="neg_root_mean_squared_error", cv=10)
pd.Series(forest_rmses).describe()

count       10.000000
mean     49684.860859
std       2034.377239
min      46989.971180
25%      47744.410115
50%      49771.948697
75%      51378.281587
max      52771.910639
dtype: float64

# Save every model you experiment with 
*using the joblib library*

In [ ]:
# CODE HERE

## now you have a shortlist of promising models. You now need to
## fine-tune them!
# Fine-Tune Your Model

## 1- Grid Search
## evaluate all the possible combinations of hyperparameter values for the RandomForestRegressor 
*It may take a long time*

In [ ]:
from sklearn.model_selection import GridSearchCV

In [ ]:
# CODE HERE

with the evaluation scores

In [ ]:
# CODE HERE

# Analyze the Best Models and Their Errors
1-indicate the relative importance of each attribute

In [ ]:
# CODE HERE

2-display these importance scores next to their corresponding attribute names:

In [ ]:
# CODE HERE

## Now is the time to evaluate the final model on the test set.
# Evaluate Your System on the Test Set

1-get the predictors and the labels from your test set

In [ ]:
# CODE HERE

2-run your full_pipeline to transform the data

In [ ]:
# CODE HERE

3-evaluate the final model on the test set

In [ ]:
# CODE HERE

# compute a 95% confidence interval for the generalization error 
*using scipy.stats.t.interval():*

In [ ]:
from scipy import stats

In [ ]:
# CODE HERE

# Great Job!
# #shAI_Club